File: backend/strategies/reversal.py
"""均值回归反转策略（RSI 超买超卖 + Bollinger %B 极值 + EMA 偏离度）。

核心思想：当价格大幅偏离均值且技术指标显示动量耗竭时，赌价格向均值回归。
与动量策略互补——动量追趋势，反转捕拐点。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from backend.strategies.momentum import (
    bollinger_at,
    bollinger_series,
    ema_series,
    rsi_series,
)


@dataclass
class ReversalParams:
    # RSI
    rsi_period: int = 14
    rsi_oversold: float = 30.0       # RSI 低于此视为超卖
    rsi_overbought: float = 70.0     # RSI 高于此视为超买
    rsi_extreme_low: float = 20.0    # 极端超卖 → 强信号
    rsi_extreme_high: float = 80.0   # 极端超买 → 强信号

    # Bollinger Band
    bb_period: int = 20
    bb_mult: float = 2.0
    pctb_low: float = 0.05           # %B 低于此 → 超卖位置
    pctb_high: float = 0.95          # %B 高于此 → 超买位置
    pctb_extreme_low: float = -0.05  # 跌出下轨 → 极端超卖
    pctb_extreme_high: float = 1.05  # 突破上轨 → 极端超买

    # EMA 偏离
    ema_period: int = 20
    deviation_entry: float = 1.5     # 偏离度 (%) 触发入场
    deviation_strong: float = 2.5    # 偏离度 (%) 触发强信号

    # 信号融合权重
    rsi_weight: float = 0.4
    bb_weight: float = 0.35
    deviation_weight: float = 0.25

    # 确认 & 冷却
    min_score: float = 0.5           # 综合得分达到此值才发出信号
    strong_score: float = 0.8        # 强信号分数线
    cooldown_bars: int = 2


def calc_reversal(vals: list[float], params: ReversalParams | None = None) -> dict[str, Any] | None:
    """
    输入收盘价序列，返回反转信号。
    返回字段与 calc_momentum 同语义：signal, strength, spreadPct 等。
    样本不足返回 None。
    """
    p = params or ReversalParams()
    min_len = max(p.rsi_period + 1, p.bb_period, p.ema_period) + 2
    if not vals or len(vals) < min_len:
        return None

    # ── RSI 分数 ───────────────────────────────────────────────
    rsi_all = rsi_series(vals, p.rsi_period)
    rsi_val = rsi_all[-1]
    rsi_score = 0.0  # -1 ~ +1，正=看多反转，负=看空反转
    if rsi_val is not None:
        if rsi_val <= p.rsi_extreme_low:
            rsi_score = 1.0
        elif rsi_val <= p.rsi_oversold:
            rsi_score = (p.rsi_oversold - rsi_val) / (p.rsi_oversold - p.rsi_extreme_low)
        elif rsi_val >= p.rsi_extreme_high:
            rsi_score = -1.0
        elif rsi_val >= p.rsi_overbought:
            rsi_score = -(rsi_val - p.rsi_overbought) / (p.rsi_extreme_high - p.rsi_overbought)

    # ── Bollinger %B 分数 ──────────────────────────────────────
    bb_now = bollinger_at(vals, p.bb_period, p.bb_mult)
    bb_score = 0.0
    bb_info: dict[str, Any] | None = None
    if bb_now:
        pctb = bb_now["percentB"]
        if pctb <= p.pctb_extreme_low:
            bb_score = 1.0
        elif pctb <= p.pctb_low:
            bb_score = (p.pctb_low - pctb) / (p.pctb_low - p.pctb_extreme_low)
        elif pctb >= p.pctb_extreme_high:
            bb_score = -1.0
        elif pctb >= p.pctb_high:
            bb_score = -(pctb - p.pctb_high) / (p.pctb_extreme_high - p.pctb_high)

        bb_prev = bollinger_at(vals[:-1], p.bb_period, p.bb_mult) if len(vals) > p.bb_period else None
        bb_info = {
            "upper": bb_now["upper"],
            "middle": bb_now["middle"],
            "lower": bb_now["lower"],
            "percentB": bb_now["percentB"],
            "bandwidth": bb_now["bandwidth"],
            "bwExpanding": bb_prev is not None and bb_now["bandwidth"] > bb_prev["bandwidth"],
        }

    # ── EMA 偏离度分数 ────────────────────────────────────────
    ema = ema_series(vals, p.ema_period)
    ema_val = ema[-1]
    price = vals[-1]
    deviation_pct = 0.0
    dev_score = 0.0
    if ema_val and ema_val > 0:
        deviation_pct = ((price - ema_val) / ema_val) * 100
        abs_dev = abs(deviation_pct)
        if abs_dev >= p.deviation_strong:
            raw = 1.0
        elif abs_dev >= p.deviation_entry:
            raw = (abs_dev - p.deviation_entry) / (p.deviation_strong - p.deviation_entry)
        else:
            raw = 0.0
        # 价格低于 EMA → 看多反转（正分），高于 EMA → 看空反转（负分）
        dev_score = raw if deviation_pct < 0 else -raw

    # ── 综合评分 ──────────────────────────────────────────────
    total_score = (
        rsi_score * p.rsi_weight
        + bb_score * p.bb_weight
        + dev_score * p.deviation_weight
    )

    abs_score = abs(total_score)
    if abs_score >= p.strong_score:
        signal = "strong_buy" if total_score > 0 else "strong_sell"
    elif abs_score >= p.min_score:
        signal = "buy" if total_score > 0 else "sell"
    else:
        signal = "neutral"

    strength = min(100.0, abs_score * 100)

    result: dict[str, Any] = {
        "signal": signal,
        "score": round(total_score, 4),
        "rsiScore": round(rsi_score, 4),
        "bbScore": round(bb_score, 4),
        "devScore": round(dev_score, 4),
        "deviationPct": round(deviation_pct, 4),
        "strength": round(strength, 2),
        "ema": round(ema_val, 4) if ema_val else None,
        "rsi": round(rsi_val, 2) if rsi_val is not None else None,
    }
    if bb_info:
        result["bb"] = {k: round(v, 4) if isinstance(v, float) else v for k, v in bb_info.items()}
    return result

File: assets/js/monitor/reversal.js
(function () {
  const Monitor = window.Monitor;
  const { app, el } = Monitor;

  // ══════════════════════════════════════════════════════════════
  // 反转策略（均值回归）—— 前端实时计算
  // ══════════════════════════════════════════════════════════════

  Monitor.reversal = {
    hu: { last: "neutral", strength: 0 },
    co: { last: "neutral", strength: 0 },
  };

  /**
   * 前端反转信号计算（与 backend/strategies/reversal.py calc_reversal 对齐）
   * @param {Array} series  [{x, y}, ...] 价格序列
   * @param {Object} params  反转参数（来自 monitor.config.json reversal 段）
   * @returns {Object|null}
   */
  Monitor.calcReversal = function (series, params, volumes) {
    const p = params || {};
    const rsiP = p.rsi_period || 14;
    const bbP = p.bb_period || 20;
    const emaP = p.ema_period || 20;
    const minLen = Math.max(rsiP + 1, bbP, emaP) + 2;
    if (!series || series.length < minLen) return null;

    const vals = series.map(v => v.y);

    // RSI 得分
    const rsiVal = Monitor.rsiAt(vals, rsiP);
    let rsiScore = 0;
    if (rsiVal != null) {
      const oversold = p.rsi_oversold || 30;
      const overbought = p.rsi_overbought || 70;
      const extLow = p.rsi_extreme_low || 20;
      const extHigh = p.rsi_extreme_high || 80;
      if (rsiVal <= extLow) rsiScore = 1;
      else if (rsiVal <= oversold) rsiScore = (oversold - rsiVal) / (oversold - extLow);
      else if (rsiVal >= extHigh) rsiScore = -1;
      else if (rsiVal >= overbought) rsiScore = -(rsiVal - overbought) / (extHigh - overbought);
    }

    // Bollinger %B 得分
    const bbMult = p.bb_mult || 2.0;
    const bbNow = Monitor.bollingerAt(vals, bbP, bbMult);
    let bbScore = 0;
    if (bbNow) {
      const pctb = bbNow.percentB;
      const pctbLow = p.pctb_low != null ? p.pctb_low : 0.05;
      const pctbHigh = p.pctb_high != null ? p.pctb_high : 0.95;
      const pctbExtLow = p.pctb_extreme_low != null ? p.pctb_extreme_low : -0.05;
      const pctbExtHigh = p.pctb_extreme_high != null ? p.pctb_extreme_high : 1.05;
      if (pctb <= pctbExtLow) bbScore = 1;
      else if (pctb <= pctbLow) bbScore = (pctbLow - pctb) / (pctbLow - pctbExtLow);
      else if (pctb >= pctbExtHigh) bbScore = -1;
      else if (pctb >= pctbHigh) bbScore = -(pctb - pctbHigh) / (pctbExtHigh - pctbHigh);
    }

    // EMA 偏离度得分
    const emaAll = Monitor.ema(vals, emaP);
    const emaVal = emaAll[emaAll.length - 1];
    const price = vals[vals.length - 1];
    let deviationPct = 0;
    let devScore = 0;
    if (emaVal && emaVal > 0) {
      deviationPct = ((price - emaVal) / emaVal) * 100;
      const devEntry = p.deviation_entry || 1.5;
      const devStrong = p.deviation_strong || 2.5;
      const absDev = Math.abs(deviationPct);
      let raw = 0;
      if (absDev >= devStrong) raw = 1;
      else if (absDev >= devEntry) raw = (absDev - devEntry) / (devStrong - devEntry);
      devScore = deviationPct < 0 ? raw : -raw;
    }

    // 成交量融合
    let volumeRatio = null;
    const volPeriod = p.volume_period || 0;
    if (volPeriod > 0 && volumes && volumes.length >= volPeriod) {
      const volEma = Monitor.ema(volumes, volPeriod);
      if (volEma && volEma.length > 0) {
        const lastVolEma = volEma[volEma.length - 1];
        if (lastVolEma != null && lastVolEma > 0) {
          const vr = volumes[volumes.length - 1] / lastVolEma;
          if (!Number.isFinite(vr) || vr < 0) return;
          volumeRatio = vr;
          let volumeScore = 0;
          const cr = p.volume_confirm_ratio != null ? p.volume_confirm_ratio : 1.5;
          const wr = p.volume_weaken_ratio != null ? p.volume_weaken_ratio : 0.6;
          if (volumeRatio > cr) {
            if (totalScore > 0) volumeScore = Math.min(1.0, (volumeRatio - 1.0) / (cr - 1.0));
            else if (totalScore < 0) volumeScore = -Math.min(1.0, (volumeRatio - 1.0) / (cr - 1.0));
          } else if (volumeRatio < wr) {
            if (totalScore > 0) volumeScore = -0.5;
            else if (totalScore < 0) volumeScore = 0.5;
          }
          totalScore += volumeScore * (p.volume_weight || 0.15);
        }
      }
    }

    // 加权综合得分
    const wRsi = p.rsi_weight || 0.4;
    const wBb = p.bb_weight || 0.35;
    const wDev = p.deviation_weight || 0.25;
    const absScore = Math.abs(totalScore);
    const minScore = p.min_score || 0.5;
    const strongScore = p.strong_score || 0.8;

    let signal = "neutral";
    if (absScore >= strongScore) signal = totalScore > 0 ? "strong_buy" : "strong_sell";
    else if (absScore >= minScore) signal = totalScore > 0 ? "buy" : "sell";

    return {
      signal,
      score: totalScore,
      rsiScore,
      bbScore,
      devScore,
      deviationPct,
      strength: Math.min(100, absScore * 100),
      rsi: rsiVal,
      bb: bbNow,
      volumeRatio,
    };
  };

  /**
   * 渲染反转信号到 DOM
   * @param {string} prefix  "hu" 或 "co"
   * @param {Object|null} info  calcReversal 返回值
   */
  Monitor.renderReversalSignal = function (prefix, info) {
    const badge = el(prefix + "RvBadge");
    const scoreEl = el(prefix + "RvScore");
    const rsiScoreEl = el(prefix + "RvRsiScore");
    const bbScoreEl = el(prefix + "RvBbScore");
    const devScoreEl = el(prefix + "RvDevScore");
    const rsiEl = el(prefix + "RvRSI");
    const devEl = el(prefix + "RvDev");
    const bar = el(prefix + "RvBar");
    const noteEl = el(prefix + "RvNote");
    const ptsEl = el(prefix + "RvPts");

    if (!badge || !bar) return;

    if (!info) {
      badge.className = "signal-badge neutral";
      badge.textContent = "等待";
      if (scoreEl) scoreEl.textContent = "--";
      if (rsiScoreEl) rsiScoreEl.textContent = "--";
      if (bbScoreEl) bbScoreEl.textContent = "--";
      if (devScoreEl) devScoreEl.textContent = "--";
      if (rsiEl) rsiEl.textContent = "--";
      if (devEl) devEl.textContent = "--";
      bar.style.width = "50%";
      bar.className = "signal-bar-inner flat";
      if (noteEl) {
        const symbol = prefix === "hu" ? "huyin" : "comex";
        const p = Monitor.getReversalParams ? Monitor.getReversalParams(symbol) : {};
        const rsiP = p.rsi_period || 14;
        const bbP = p.bb_period || 20;
        const emaP = p.ema_period || 20;
        const minLen = Math.max(rsiP + 1, bbP, emaP) + 2;
        noteEl.textContent = `需至少 ${minLen} 个有效价格点后计算`;
      }
      if (ptsEl) ptsEl.textContent = "--";
      return;
    }

    badge.className = "signal-badge " + info.signal.replace("_", "-");
    badge.textContent = Monitor.signalLabels[info.signal];

    const fmtScore = (v) => {
      const cls = v > 0 ? "up" : v < 0 ? "down" : "";
      return `<span class="val ${cls}">${v >= 0 ? "+" : ""}${v.toFixed(3)}</span>`;
    };

    if (scoreEl) scoreEl.innerHTML = fmtScore(info.score);
    if (rsiScoreEl) rsiScoreEl.innerHTML = fmtScore(info.rsiScore);
    if (bbScoreEl) bbScoreEl.innerHTML = fmtScore(info.bbScore);
    if (devScoreEl) devScoreEl.innerHTML = fmtScore(info.devScore);

    if (rsiEl && info.rsi != null) {
      const r = info.rsi;
      const cls = r > 70 ? "up" : r < 30 ? "down" : "";
      rsiEl.innerHTML = `<span class="val ${cls}">${r.toFixed(1)}</span>`;
    } else if (rsiEl) {
      rsiEl.textContent = "--";
    }

    if (devEl) {
      const d = info.deviationPct;
      const cls = d > 0 ? "up" : d < 0 ? "down" : "";
      devEl.innerHTML = `<span class="val ${cls}">${d >= 0 ? "+" : ""}${d.toFixed(3)}%</span>`;
    }

    // 成交量比率渲染
    const volEl = el(prefix + "RvVolRatio");
    if (volEl && info.volumeRatio != null) {
      const vr = info.volumeRatio;
      const vcls = vr > 1.5 ? "up" : vr < 0.6 ? "down" : "";
      volEl.innerHTML = `<span class="val ${vcls}">${vr.toFixed(2)}x</span>`;
    } else if (volEl) {
      volEl.textContent = "--";
    }

    bar.style.width = info.strength.toFixed(0) + "%";
    bar.className = "signal-bar-inner " + (info.signal.includes("buy") ? "bull" : info.signal.includes("sell") ? "bear" : "flat");

    if (ptsEl) ptsEl.textContent = "实时";

    if (noteEl) {
      let note = `综合 ${info.score >= 0 ? "+" : ""}${info.score.toFixed(3)}`;
      if (info.rsi != null) {
        if (info.rsi < 30) note += " | RSI超卖";
        else if (info.rsi > 70) note += " | RSI超买";
      }
      if (info.bb) {
        if (info.bb.percentB < 0) note += " | 跌破下轨";
        else if (info.bb.percentB > 1) note += " | 突破上轨";
      }
      if (Math.abs(info.deviationPct) > 1.5) note += " | 偏离均值";
      noteEl.textContent = note;
    }
  };

  Monitor.updateReversalSignals = function (data) {
    // 后端可能通过 /api/all 返回或 SSE 推送 precomputed reversal signals
    const backendRv = data.reversalSignals || data.reversal_signals || Monitor.instrumentReversalSignals || {};
    const rtBuffers = data.realtimeBacktestBuffers || {};

    // fallback 时也优先使用后端 realtimeBacktestBuffers（1秒采样），确保前后端使用同一序列
    const _rtSeries = (buf) => (buf && buf.length >= 2) ? buf.map(p => ({ x: p.t, y: p.y })) : null;
    const _rtVolumes = (buf) => (buf && buf.length >= 2) ? buf.map(p => p.v).filter(v => v != null) : [];
    const huSeries = _rtSeries(rtBuffers.ag0) || app.silverLivePoints.map(p => ({ x: p.t, y: p.y }));
    const huVolumes = _rtVolumes(rtBuffers.ag0).length > 0 ? _rtVolumes(rtBuffers.ag0) : app.silverLivePoints.map(p => p.v).filter(v => v != null);
    const coSeries = _rtSeries(rtBuffers.xag) || app.comexSilverLivePoints.map(p => ({ x: p.t, y: p.y }));
    const coVolumes = _rtVolumes(rtBuffers.xag).length > 0 ? _rtVolumes(rtBuffers.xag) : app.comexSilverLivePoints.map(p => p.v).filter(v => v != null);

    const huParams = Monitor.getReversalParams("huyin");
    const coParams = Monitor.getReversalParams("comex");

    // 有效性检查：后端信号需包含至少 signal 或 rsi 等字段
    const _valid = s => s && (s.signal != null || s.score != null || s.rsi != null);

    const huSig = _valid(backendRv.ag0) ? backendRv.ag0 : Monitor.calcReversal(huSeries, huParams, huVolumes);
    const coSig = _valid(backendRv.xag) ? backendRv.xag : Monitor.calcReversal(coSeries, coParams, coVolumes);

    if (huSig && Monitor.reversal.hu.last !== huSig.signal) {
      Monitor.playSignalTone(huSig.signal);
      Monitor.reversal.hu = { last: huSig.signal, strength: huSig.strength };
    }
    if (coSig && Monitor.reversal.co.last !== coSig.signal) {
      Monitor.playSignalTone(coSig.signal);
      Monitor.reversal.co = { last: coSig.signal, strength: coSig.strength };
    }

    Monitor.renderReversalSignal("hu", huSig);
    Monitor.renderReversalSignal("co", coSig);
  };

})();
